### PR TITLE
[Web/JS] Add ConvTranspose support

### PR DESCRIPTION
**Pull request by satyajandhyala**
### Description
Add ConvTranspose support for WebGPU

### CODE REVIEW

File: js/web/lib/onnxjs/attribute.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ import ortFbs = onnxruntime.experimental.fbs;
 
 export declare namespace Attribute {
   export interface DataTypeMap {
+    boolean: boolean;
```
Comment (fs-eire) [inline js/web/lib/onnxjs/attribute.ts:15]: I think this change is incorrect. Firstly this file is not related to the PR ( it's in onnxjs/ folder and should have nothing to do with webgpu ) , and secondly in onnx spec we don't have boolean attribute.

File: js/web/lib/wasm/jsep/webgpu/ops/conv-transpose.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+          convInputs.push(inputs[2].reshape([inputs[2].dims[0], 1, 1]));
+        } else {
+          convInputs.push(inputs[2]);
+        }
+      }
+
+      // STEP.3: compute matmul
+      context.compute(
+          createConvTranspose2DProgramInfoLoader(convInputs, adjustedAttributes, outputShape, hasBias),
+          {inputs: convInputs});
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/conv-transpose.ts:251]: the current implementation depends on a pre-transposed weight, which is not "naive". A native is expected to be as simple as possible ( mostly is a brute-force implementation ).

and the file organization is confusing. the current files look like there are 2 different implementations - one is an optimized on defined in conv-transpose.ts and the other is a simple one defined in conv-transpose-naive.ts. it turns out that there is only one implementation. I suggest to merge those 2 files into one unless we really get 2 different implementations.

File: onnxruntime/core/providers/js/operators/conv_transpose.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      }
+      for (size_t i = 0; i < conv_transpose_attrs_.dilations.size() && i < dialations_vec_size; ++i) {
+        local_dilations[i] = conv_transpose_attrs_.dilations[i];
+      }
+      for (size_t i = 0; i < conv_transpose_attrs_.strides.size() && i < strides_vec_size; ++i) {
+        local_strides[i] = conv_transpose_attrs_.strides[i];
+      }
+      LOGS_DEFAULT(VERBOSE) << "output_shape = " << conv_transpose_attrs_.output_shape << std::endl;
+      LOGS_DEFAULT(VERBOSE) << "output_padding = " << conv_transpose_attrs_.output_padding << std::endl;
+      JSEP_INIT_KERNEL_ATTRIBUTE(ConvTranspose, ({
```
Comment (fs-eire) [inline onnxruntime/core/providers/js/operators/conv_transpose.h:84]: I feel sorry to make you have to write those. I am thinking about how to do this more simple

File: js/web/lib/wasm/jsep/webgpu/ops/3rd-party/conv_backprop_webgpu.ts [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+            if (bDyCVal && bDyCVal2) {
+              let d2Length = outBackprop[3];
+              for (var d2 :u32 = 0; d2 < d2Length; d2 = d2 + 4) {
+                ${
+          wIndicesHelper.indicesVariableDeclaration(
+              'wIndices0',
+              isChannelsLast ? ['d2', 'wRPerm', 'wCPerm', 'd1'] :
+                               [
+                                 'd2', 'd1', 'wRPerm', 'wCPerm'
+                               ])};
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/3rd-party/conv_backprop_webgpu.ts:110]: I assume that `isChannelsLast` is for X (input[0]), not for W (input[1]). And for W it should always be channel last , because you always transpose it in conv-transpose.ts. if you don't tranpose, it should be always align with spec https://onnx.ai/onnx/operators/onnx__ConvTranspose.html (C x M/group x kH x kW)

File: js/web/lib/wasm/jsep/webgpu/ops/conv-transpose.ts (hunk 2 above)
Comment (satyajandhyala) [reply js/web/lib/wasm/jsep/webgpu/ops/conv-transpose.ts:251]: I have created conv-transpose-naive.ts only as a wrapper to the implementation in 3rd-pary/conv-backprop-webgpu.ts along the same lines as conv2d-mm.ts. Does changing the name from conv-transpose-naive.ts to conv-transpose2d-naive.ts make it clear? My idea is that we eventually add a MatMul version also.

**Review by fs-eire — APPROVED** [review]
